github/downloader.go

```diff
@@ -0,0 +1,921 @@
+	process := func(review *graphql.PullRequestReview) error {
+		err := d.storer.SavePullRequestReview(owner, name, pr.Number, review)
+		if err != nil {
+			return fmt.Errorf("failed to save PR review for PR %v/%v #%v: %v", owner, name, pr.Number, err)
+		}
+		return d.downloadReviewComments(owner, name, pr.Number, review)
+	}
+
+	// save first page of reviews
+	for _, review := range pr.Reviews.Nodes {
+		err := process(&review)
+		if err != nil {
+			return err
+		}
+	}
+
+	variables := map[string]interface{}{
+		"owner":    githubv4.String(owner),
+		"name":     githubv4.String(name),
+		"prNumber": githubv4.Int(pr.Number),
+
+		"pullRequestReviewCommentsPage": githubv4.Int(pullRequestReviewCommentsPage),
+		"pullRequestReviewsPage":        githubv4.Int(pullRequestReviewsPage),
+
+		"pullRequestReviewCommentsCursor": (*githubv4.String)(nil),
+		"pullRequestReviewsCursor":        (*githubv4.String)(nil),
+	}
+
+	// if there are more reviews, loop over all the pages
+	hasNextPage := pr.Reviews.PageInfo.HasNextPage
+	endCursor := pr.Reviews.PageInfo.EndCursor
+
+	for hasNextPage {
+		// get only PR reviews
+		var q struct {
+			Repository struct {
+				PullRequest struct {
+					Reviews graphql.PullRequestReviewConnection `graphql:"reviews(first: $pullRequestReviewsPage, after: $pullRequestReviewsCursor)"`
+				} `graphql:"pullRequest(number: $prNumber)"`
+			} `graphql:"repository(owner: $owner, name: $name)"`
+		}
+
+		variables["pullRequestReviewsCursor"] = githubv4.String(endCursor)
+
+		err := d.client.Query(ctx, &q, variables)
+		if err != nil {
+			return fmt.Errorf("failed to query PR reviews for PR %v/%v #%v: %v", owner, name, pr.Number, err)
+		}
+
+		for _, review := range q.Repository.PullRequest.Reviews.Nodes {
+			process(&review)
+		}
+
+		hasNextPage = q.Repository.PullRequest.Reviews.PageInfo.HasNextPage
+		endCursor = q.Repository.PullRequest.Reviews.PageInfo.EndCursor
+	}
+
+	return nil
+}
+
+func (d Downloader) downloadReviewComments(repositoryOwner, repositoryName string, pullRequestNumber int, review *graphql.PullRequestReview) error {
+	// save first page of comments
+	for _, comment := range review.Comments.Nodes {
+		err := d.storer.SavePullRequestReviewComment(repositoryOwner, repositoryName, pullRequestNumber, review.DatabaseId, &comment)
+		if err != nil {
+			return fmt.Errorf("failed to save PR review comment for PR %v/%v #%v: %v", repositoryOwner, repositoryName, pullRequestNumber, err)
+		}
+	}
+
+	/*
+		TODO: how to perform pagination at this level
+
+		There isn't a way to ask for repository/pullRequest(number:3)/review(id:X),
+		you can only query all reviews
+
+		{
+		  repository(owner: "carlosms-test-org", name: "test-repo") {
+		    pullRequest(number:3) {
+		      reviews {.....}
+		    }
+		  }
+		}
+	*/
+	if review.Comments.PageInfo.HasNextPage {
+		log.Errorf(nil,
+			"%v/%v PR #%v review %v contains more than one page of comments (%v), but pagination is not implemented",
```

Comment:
oh. now I see where the error comes from.

Reply:
This is now implemented in 5f472bd